smart_contracts/ethereum/solidity_firefly/contracts/Firefly.sol

```diff
@@ -2,27 +2,39 @@
 
 pragma solidity >=0.6.0 <0.9.0;
 
-contract Firefly {
-    event BatchPin(
-        address author,
-        uint timestamp,
-        string action,
-        bytes32 uuids,
-        bytes32 batchHash,
-        string payloadRef,
-        bytes32[] contexts
-    );
+import "./IBatchPin.sol";
+
+contract Firefly is IBatchPin {
+    function pinBatchData(bytes calldata data) public override {
+        bytes32 uuids;
+        bytes32 batchHash;
+        string memory payloadRef;
+        bytes32[] memory contexts;
+        (uuids, batchHash, payloadRef, contexts) = abi.decode(
+            data,
+            (bytes32, bytes32, string, bytes32[])
+        );
+        emit BatchPin(
+            tx.origin,
+            block.timestamp,
+            "firefly:contract_invoke_pin",
```

Comment:
Like the clarity of this eyecatcher.
Good to keep it under 32b length string (this is 28), as that's 2-words (len+data) in Eth storage.